File: Laya/GPUSkinning/src/GPUSkinning/Datas/GPUSkinningFrame.ts
import Matrix4x4 = Laya.Matrix4x4;
import Quaternion = Laya.Quaternion;
import ByteReadUtil from "./ByteReadUtil";

/** 烘焙动画--动画剪辑帧数据信息 */
export default class GPUSkinningFrame
{
    /** 矩阵 */
    matrices: Matrix4x4[];

    /** 相对根节点偏移--方向  */
    rootMotionDeltaPositionQ: Quaternion;
    
    /** 相对根节点偏移--距离  */
    rootMotionDeltaPositionL: float;

    /** 相对根节点偏移 */
    rootMotionDeltaRotation: Quaternion;

    /** 是否根节点初始化逆矩阵 */
    private rootMotionInvInit: boolean = false;
    private rootMotionInv: Matrix4x4;
    /** 根节点逆矩阵 */
    public RootMotionInv(rootBoneIndex: int ):Matrix4x4
    {
        if (!this.rootMotionInvInit)
        {
            let m =  this.matrices[rootBoneIndex];
            this.rootMotionInv = new Matrix4x4();
            m.invert(this.rootMotionInv);
            this.rootMotionInvInit = true;
        }
        return this.rootMotionInv;
    }

    
    FromBytes(data: ArrayBuffer, rootMotionEnabled:boolean): void
    {
        var b:Laya.Byte = new Laya.Byte(data);
        if(rootMotionEnabled)
        {
            this.rootMotionDeltaPositionL = b.readFloat32();
            this.rootMotionDeltaPositionQ = ByteReadUtil.ReadQuaternion(b);
            this.rootMotionDeltaRotation = ByteReadUtil.ReadQuaternion(b);
        }

        var matricesCount = b.readUint32();
        this.matrices = [];
        for(var i = 0; i < matricesCount; i ++)
        {
            var m = ByteReadUtil.ReadMatrix4x4(b);
            this.matrices.push(m);
        }
        

    }

    static CreateFromBytes(data: ArrayBuffer, rootMotionEnabled:boolean):GPUSkinningFrame
    {
        var obj = new GPUSkinningFrame();
        obj.FromBytes(data, rootMotionEnabled);
        return obj;
    }



}

File: Unity3D/GPUSkinning/Assets/GPUSkinning/Scripts/GPUSkinningFrame.cs
using UnityEngine;
using System.Collections;
using System.IO;
using System.Collections.Generic;

[System.Serializable]
public class GPUSkinningFrame
{
    public Matrix4x4[] matrices = null;

    public Quaternion rootMotionDeltaPositionQ;

    public float rootMotionDeltaPositionL;

    public Quaternion rootMotionDeltaRotation;

    [System.NonSerialized]
    private bool rootMotionInvInit = false;
    [System.NonSerialized]
    private Matrix4x4 rootMotionInv;
    public Matrix4x4 RootMotionInv(int rootBoneIndex)
    {
        if (!rootMotionInvInit)
        {
            rootMotionInv = matrices[rootBoneIndex].inverse;
            rootMotionInvInit = true;
        }
        return rootMotionInv;
    }


    public static GPUSkinningFrame CreateFromBytes(byte[] bytes)
    {
        GPUSkinningFrame obj = new GPUSkinningFrame();
        obj.FromBytes(bytes);
        return obj;
    }

    public void FromBytes(byte[] bytes)
    {
    }


    public MemoryStream ToSteam(List<int> exportBoneIndexList,  bool rootMotionEnabled)
    {
        MemoryStream stream = new MemoryStream();
        BinaryWriter b = new BinaryWriter(stream);

        if(rootMotionEnabled)
        {
            b.Write((float)rootMotionDeltaPositionL);
            b.WriteQuaternion(rootMotionDeltaPositionQ);
            b.WriteQuaternion(rootMotionDeltaRotation);
        }



        // 矩阵列表 数量
        b.Write((uint)exportBoneIndexList.Count);
        foreach(int index in exportBoneIndexList)
        {
            b.WriteMatrix4x4(matrices[index]);
        }



        return stream;
    }
}
